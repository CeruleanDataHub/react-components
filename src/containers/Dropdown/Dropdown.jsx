import PropTypes from "prop-types";
import React, { useState } from "react";
import styled from "styled-components";

import { Button } from "../Button/Button";
import { Checkbox } from "../Checkbox/Checkbox";
import { Icon } from "../Icon/Icon";
import { Typography } from "../Typography/Typography";

const Container = styled.div`
  position: relative;
`;

const DropdownContainer = styled.button`
  font-family: inherit;
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  height: 40px;
  border: 1px solid ${({ theme }) => theme.border};
  background: ${({ theme }) => theme.background};
  border-radius: 4px;
  padding: 1.25em;
  cursor: pointer;
`;

const ItemList = styled.ul`
  width: 100%;
  padding: 0;
  background: ${({ theme }) => theme.background};
  border: 1px solid ${({ theme }) => theme.border};
  border-top: none;
  position: absolute;
  top: 22px;
  border-radius: 0 0 4px 4px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  list-style: none;
  z-index: 1;
  box-sizing: border-box;
`;

const DropdownItem = styled.li`
  padding-top: 1em;
  padding-left: 0.5em;
`;

const DropdownText = styled.span`
  color: ${({ green, theme }) => (green ? theme.green : "#999")};
  font-size: 1.25em;
  font-weight: bold;
`;

const generateList = (items, handleSelection) =>
  items.map(({ value, label, checked }) => (
    <DropdownItem key={label}>
      <Checkbox
        checked={checked}
        value={value}
        label={label}
        onChange={handleSelection}
      />
    </DropdownItem>
  ));

const setSelectedItemAsChecked = value => items =>
  items.map(item => ({
    ...item,
    checked: item.value === value ? !item.checked : item.checked
  }));

export const Dropdown = ({ isOpen, label, onClick, items }) => {
  const [selectedItems, setSelectedItems] = useState(
    items.map(item => ({ ...item, checked: false }))
  );

  const checkedItems = selectedItems.filter(({ checked }) => checked);

  const handleSelection = value => {
    setSelectedItems(setSelectedItemAsChecked(value));
  };

  return (
    <Typography fontFamily="openSans">
      <Container>
        <Button onClick={onClick} as={DropdownContainer}>
          <DropdownText>{label}</DropdownText>
          {isOpen && (
            <DropdownText green>
              {`${checkedItems.length} selected`}
            </DropdownText>
          )}
          <Icon
            name={isOpen ? "chevron-up" : "chevron-down"}
            as={DropdownText}
          />
        </Button>
        {isOpen && (
          <ItemList>{generateList(selectedItems, handleSelection)}</ItemList>
        )}
      </Container>
    </Typography>
  );
};

Dropdown.propTypes = {
  /** Define dropdown visibility state */
  isOpen: PropTypes.bool,
  /** Text to show in dropdown */
  label: PropTypes.string,
  /** Handler function */
  onClick: PropTypes.func,
  /** Items to be shown in dropdown list */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string
    })
  )
};

Dropdown.defaultProps = {
  isOpen: false,
  label: "",
  onClick: () => {},
  items: []
};
