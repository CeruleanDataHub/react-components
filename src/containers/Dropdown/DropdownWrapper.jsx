import PropTypes from "prop-types";
import React, { useState } from "react";

import { Dropdown } from "./Dropdown";

export const DropdownWrapper = ({ onClick, label, items }) => {
  const [isOpen, setIsOpen] = useState(false);

  const onDropdownClick = () => {
    setIsOpen(!isOpen);
    onClick(isOpen);
  };

  return (
    <Dropdown
      label={label}
      onClick={onDropdownClick}
      isOpen={isOpen}
      items={items}
    />
  );
};

DropdownWrapper.propTypes = {
  onClick: PropTypes.func,
  label: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string
    })
  )
};

DropdownWrapper.defaultProps = {
  onClick: () => {},
  label: "",
  items: []
};
