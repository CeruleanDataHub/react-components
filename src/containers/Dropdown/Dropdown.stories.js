import { action } from "@storybook/addon-actions";
import React from "react";

import { Dropdown } from "./Dropdown";
import { DropdownWrapper } from "./DropdownWrapper";

export default {
  title: "Dropdown",
  component: Dropdown
};

const items = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" },
  { value: "3", label: "Item 3" },
  { value: "4", label: "Item 4" },
  { value: "5", label: "Item 5" }
];

export const DropdownStory = () => <Dropdown label="Label" />;

export const DropdownOpenStory = () => (
  <div style={{ minHeight: 250 }}>
    <Dropdown label="Label" isOpen items={items} />
  </div>
);

export const DropdownToggleStory = () => (
  <div style={{ minHeight: 250 }}>
    <DropdownWrapper
      label="Label"
      items={items}
      onClick={action("dropdown-open")}
    />
  </div>
);
