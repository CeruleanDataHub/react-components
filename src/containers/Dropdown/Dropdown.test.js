import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import renderer from "react-test-renderer";

import { Dropdown } from "./Dropdown";

const items = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" }
];

describe("Dropdown", () => {
  it("should render dropdown", () => {
    const component = renderer.create(<Dropdown />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render dropdown in open state", () => {
    const component = renderer.create(<Dropdown items={items} isOpen />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should have onClick listener", () => {
    const component = renderer.create(
      <Dropdown items={items} onClick={() => null} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should fire onClick event callback function", () => {
    const handleClick = jest.fn();
    const component = mount(<Dropdown items={items} onClick={handleClick} />);
    expect(handleClick).not.toHaveBeenCalled();
    component.find("button").simulate("click");
    expect(handleClick).toHaveBeenCalled();
  });

  it("should call default props onClick function when no onClick property is passed", () => {
    const component = renderer.create(<Dropdown items={items} />);
    const tree = component.toJSON();
    component.root.props.onClick();
    expect(tree).toMatchSnapshot();
  });

  it("should be able to select item from list", () => {
    const wrapper = mount(<Dropdown items={items} isOpen />);
    wrapper
      .find("input")
      .first()
      .simulate("change");

    expect(toJson(wrapper, { mode: "deep" })).toMatchSnapshot();
  });
});
