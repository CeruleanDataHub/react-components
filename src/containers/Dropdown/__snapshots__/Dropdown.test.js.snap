// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render dropdown 1`] = `
.c3 {
  font-family: inherit;
  background: transparent;
  font-size: 1em;
  padding: 0.25em 1.5em;
  border: 1px solid;
  border-radius: 4px;
  cursor: pointer;
}

.c5 {
  font-size: 4rem;
  text-align: center;
}

.c0 {
  font-family: "Open Sans",sans-serif;
  font-size: 16px;
  color: black;
}

.c1 {
  position: relative;
}

.c2 {
  font-family: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 40px;
  border: 1px solid #0f181c;
  background: #0f181c;
  border-radius: 4px;
  padding: 1.25em;
  cursor: pointer;
}

.c4 {
  color: #999;
  font-size: 1.25em;
  font-weight: bold;
}

<div
  className="c0"
  fontFamily="openSans"
  size="normal"
>
  <div
    className="c1"
  >
    <button
      className="c2 c3"
      color=""
      onClick={[Function]}
    >
      <span
        className="c4"
      >
        
      </span>
      <span
        className="c4 c5 lni lni-chevron-down"
      />
    </button>
  </div>
</div>
`;

exports[`Dropdown should render dropdown in open state 1`] = `
.c3 {
  font-family: inherit;
  background: transparent;
  font-size: 1em;
  padding: 0.25em 1.5em;
  border: 1px solid;
  border-radius: 4px;
  cursor: pointer;
}

.c6 {
  font-size: 4rem;
  text-align: center;
}

.c10 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c12 {
  display: inline-block;
  width: 1em;
  height: 1em;
  border: 1px solid #999;
  background: transparent;
  text-align: center;
  border-radius: 4px;
  line-height: 1em;
}

.c14 {
  color: #999;
  margin-left: 1.5em;
}

.c9 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
  margin: 0.25em 0;
}

.c9:hover .c11 {
  border-color: #333;
}

.c9:hover .c13 {
  color: #666;
}

.c0 {
  font-family: "Open Sans",sans-serif;
  font-size: 16px;
  color: black;
}

.c1 {
  position: relative;
}

.c2 {
  font-family: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 40px;
  border: 1px solid #0f181c;
  background: #0f181c;
  border-radius: 4px;
  padding: 1.25em;
  cursor: pointer;
}

.c7 {
  width: 100%;
  padding: 0;
  background: #0f181c;
  position: absolute;
  top: 22px;
  border-radius: 0 0 4px 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  list-style: none;
  z-index: 1;
}

.c8 {
  padding-top: 1em;
  padding-left: 0.5em;
}

.c4 {
  color: #999;
  font-size: 1.25em;
  font-weight: bold;
}

.c5 {
  color: #90ee7e;
  font-size: 1.25em;
  font-weight: bold;
}

<div
  className="c0"
  fontFamily="openSans"
  size="normal"
>
  <div
    className="c1"
  >
    <button
      className="c2 c3"
      color=""
      onClick={[Function]}
    >
      <span
        className="c4"
      >
        
      </span>
      <span
        className="c5"
        color="#90ee7e"
      >
        0 selected
      </span>
      <span
        className="c4 c6 lni lni-chevron-up"
      />
    </button>
    <ul
      className="c7"
    >
      <li
        className="c8"
      >
        <label
          className="c9"
        >
          <input
            checked={false}
            className="c10"
            onChange={[Function]}
            type="checkbox"
            value="1"
          />
          <div
            checked={false}
            className="c11 c12"
          />
          <span
            className="c13 c14"
          >
            Item 1
          </span>
        </label>
      </li>
      <li
        className="c8"
      >
        <label
          className="c9"
        >
          <input
            checked={false}
            className="c10"
            onChange={[Function]}
            type="checkbox"
            value="2"
          />
          <div
            checked={false}
            className="c11 c12"
          />
          <span
            className="c13 c14"
          >
            Item 2
          </span>
        </label>
      </li>
    </ul>
  </div>
</div>
`;
